### PR TITLE
syntax: Rewrite parsing of patterns

### LINKED ISSUE #22546: add more tests for type-instantiation `pat::<T>` syntax in patterns
Spawned off of issue #22544 

Consider this code:

``` rust
pub struct Foo<T>(T, T);

impl<T: ::std::fmt::Display> Foo<T> {
    fn foo(&self) {
        match *self {
            Foo::<T>(ref x, ref y) =>  // <=== check out this line!
                println!("Goodbye, World! {} {}", x, y)
        }
    }
}

fn main() {
    let f = Foo(2,3);
    f.foo();
}
```

The above compiles and runs in the playpen today.

We probably shouldn't accept the use of `Foo::<T>` in a pattern, right?

(Is there some context where we _would_ want that, i.e. as some sort of feedback into type-inference?  I dunno, I still remain to be convinced that type ascription in patterns is generally useful.)

### LINKED ISSUE #22757: ICE with a simple bintree, compiler panicked
```rust
#[derive(Debug)]
enum BtNode {
    Node(u32,Box<BtNode>,Box<BtNode>),
    Leaf(u32),
}

impl BtNode {
    fn stringify(&self) -> String {
        match *self {
            BtNode::Leaf(val) => { format!("Leaf: {}",val) }
            BtNode::Node(val,ref l,ref r) => {format!("Node: {} {:?} {:?}", val,l,r) }
        }
    }
    fn setLeft(&self, Box<BtNode> left) -> &BtNode {
        match *self {
            BtNode::Leaf(val) => { }
            BtNode::Node(val,ref l,ref r) => { l = left }
        }
        return self;
    }
}

fn main() {
    let node =  Box::new( BtNode::Node( 1, Box::new(BtNode::Leaf(0)), Box::new(BtNode::Leaf(0)) )) ;
    let node2 = Box::new( BtNode::Node( 2, Box::new(BtNode::Leaf(0)), Box::new(BtNode::Leaf(0)) )) ;
    let root = Box::new( BtNode::Node( 0, node, node2 ) );
}
```

`14:26 error: internal compiler error: ident only path should have been covered already`

### LINKED ISSUE #22972: internal compiler error: ident only path should have been covered already
```
$ cat testje.rs
fn main(){
    let mut v=std::vec::Vec::<isize>::new();
    v.push(1);
    v.push(2);
    // The next line makes rust crash.  It does not make any sense (it was just a typo) but thought i'd report it anyway.
    let v[0]=v[1];
}

$ rustc --version
rustc 1.0.0-nightly (890293655 2015-02-28) (built 2015-03-01)

$ RUST_BACKTRACE=1 rustc testje.rs
testje.rs:6:9: 6:10 error: internal compiler error: ident only path should have been covered already
testje.rs:6     let v[0]=v[1];
                    ^
note: the compiler unexpectedly panicked. this is a bug.
note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports
note: run with `RUST_BACKTRACE=1` for a backtrace
thread 'rustc' panicked at 'Box<Any>', /home/rustbuild/src/rust-buildbot/slave/nightly-dist-rustc-linux/build/src/libsyntax/diagnostic.rs:129

stack backtrace:
   1:         0x577a6c4f - sys::backtrace::write::hb08af03691eab78a5kA
   2:         0x577cf782 - panicking::on_panic::h37164609ca1c02afBnJ
   3:         0x5770f63a - rt::unwind::begin_unwind_inner::h24d5e7492198457b43I
   4:         0x54a55a8d - rt::unwind::begin_unwind::h9113908695753286194
   5:         0x54a55a33 - diagnostic::SpanHandler::span_bug::hcbc27df468564853WBD
   6:         0x54a7b905 - parse::parser::Parser<'a>::parse_pat::hf7a2ebdbbed007b1d3J
   7:         0x54a8df2d - parse::parser::Parser<'a>::parse_local::h7d6c62c9c63efba1xkK
   8:         0x54a8f22f - parse::parser::Parser<'a>::parse_stmt::h622ba475055891c9inK
   9:         0x54a97e89 - parse::parser::Parser<'a>::parse_block_tail_::hb2f43cc30c93d511ryK
  10:         0x54a75dec - parse::parser::Parser<'a>::parse_inner_attrs_and_block::h554d87549b4ccdacTwK
  11:         0x54a9c465 - parse::parser::Parser<'a>::parse_item_fn::hbda705c8fc5a7ac1icL
  12:         0x54a92778 - parse::parser::Parser<'a>::parse_item_::hf9f9cf23aa0f1cd4iaM
  13:         0x54aa2985 - parse::parser::Parser<'a>::parse_mod_items::habe606a27fa5b86dSAL
  14:         0x54aacc6c - parse::parser::Parser<'a>::parse_crate_mod::h8c08698d71a0eba6PAM
  15:         0x54abf61e - parse::parse_crate_from_file::ha077e55f357e36faJNT
  16:         0x57e341b1 - driver::phase_1_parse_input::closure.15399
  17:         0x57e0e7b7 - driver::phase_1_parse_input::h21d9f24721445c1fvqa
  18:         0x57e0b181 - driver::compile_input::hee81c045ab46c74fIba
  19:         0x57ed6e6e - run_compiler::h875c1e123fe4ae2cF5b
  20:         0x57ed4a1c - thunk::F.Invoke<A, R>::invoke::h12424714292252424185
  21:         0x57ed3770 - rt::unwind::try::try_fn::h4021821266450575721
  22:         0x5783afe8 - rust_try_inner
  23:         0x5783afd5 - rust_try
  24:         0x57ed3e5c - thunk::F.Invoke<A, R>::invoke::h8199562432922027576
  25:         0x577bb2d5 - sys::thread::thread_start::h1b1bb1e7ac08d3a6yRE
  26:         0x5150db4f - start_thread
  27:         0x573a270c - <unknown>
```

There are a few issues with the same error message, but they seem to have been fixed before my nightly  - and the way to cause it is also different

### LINKED ISSUE #23044: ICE: ident only path for external function declaration
```
rustc 1.0.0-dev (f0f7ca27d 2015-02-23) (built 2015-02-23)
```
```
building application
   Compiling android_glue v0.0.1 (file:///mnt/host/kmcg3413.net/FireRoute/android-clui)
/mnt/host/kmcg3413.net/FireRoute/apk-builder/glue/src/lowglue/mod.rs:39:25: 39:28 error: internal compiler error: ident only path should have been covered already
/mnt/host/kmcg3413.net/FireRoute/apk-builder/glue/src/lowglue/mod.rs:39     pub fn android_main(Arc<ActivityState>);
                                                                                                ^~~
note: the compiler unexpectedly panicked. this is a bug.
note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports
note: run with `RUST_BACKTRACE=1` for a backtrace
thread 'rustc' panicked at 'Box<Any>', /mnt/host/kmcg3413.net/rust/src/libsyntax/diagnostic.rs:129

stack backtrace:
   1:     0x7fd0ca974090 - sys::backtrace::write::h4e9a31aa2e8023179rC
   2:     0x7fd0ca99cca0 - panicking::on_panic::hc7167c0722a05a78b9L
   3:     0x7fd0ca8dc860 - rt::unwind::begin_unwind_inner::h8d46481699e3751fwNL
   4:     0x7fd0c7bd5430 - rt::unwind::begin_unwind::h14067791790174494748
   5:     0x7fd0c7bd53c0 - diagnostic::SpanHandler::span_bug::h85d108ba45cb62f281E
   6:     0x7fd0c7bf7e70 - parse::parser::Parser<'a>::parse_pat::h6d0073e3d8694344ZPL
   7:     0x7fd0c7bf08b0 - parse::parser::Parser<'a>::parse_arg_general::h8114594c0bc8a45ei9J
   8:     0x7fd0c7c1bde0 - parse::parser::Parser<'a>::parse_seq_to_before_end::h1598884596488876550
   9:     0x7fd0c7bef6c0 - parse::parser::Parser<'a>::parse_fn_args::h47c7163922b1aa1b1EM
  10:     0x7fd0c7c1c470 - parse::parser::Parser<'a>::parse_fn_decl::h2b34c90f31c289d7wHM
  11:     0x7fd0c7c2d020 - parse::parser::Parser<'a>::parse_foreign_item::hb5b9d02987c373a3rdO
  12:     0x7fd0c7c29880 - parse::parser::Parser<'a>::parse_foreign_mod_items::h608317c1fc7b75890JN
  13:     0x7fd0c7c11540 - parse::parser::Parser<'a>::parse_item_::h5342af3cfac016bc3YN
  14:     0x7fd0c7c22ab0 - parse::parser::Parser<'a>::parse_mod_items::hcf50c5d5eff08ba3ioN
  15:     0x7fd0c7c23e40 - parse::parser::Parser<'a>::eval_src_mod::hefec9279931acc91iwN
  16:     0x7fd0c7c11540 - parse::parser::Parser<'a>::parse_item_::h5342af3cfac016bc3YN
  17:     0x7fd0c7c22ab0 - parse::parser::Parser<'a>::parse_mod_items::hcf50c5d5eff08ba3ioN
  18:     0x7fd0c7c2dfe0 - parse::parser::Parser<'a>::parse_crate_mod::h3a386f4cf831334c9pO
  19:     0x7fd0c7c40d30 - parse::parse_crate_from_file::he93fe4d172bcd8ccjSV
  20:     0x7fd0caffe510 - driver::phase_1_parse_input::closure.15704
  21:     0x7fd0cafd8840 - driver::phase_1_parse_input::h4fb166827a2f9cbcYra
  22:     0x7fd0cafd5240 - driver::compile_input::h0bec8d7ff7428d7dGba
  23:     0x7fd0cb0a4330 - run_compiler::h71833f9bcc896daaWbc
  24:     0x7fd0cb0a2c30 - thunk::F.Invoke<A, R>::invoke::h14347372239080705007
  25:     0x7fd0cb0a1b20 - rt::unwind::try::try_fn::h10841254516551992180
  26:     0x7fd0caa0a280 - rust_try_inner
  27:     0x7fd0caa0a270 - rust_try
  28:     0x7fd0cb0a1e20 - thunk::F.Invoke<A, R>::invoke::h15709534366616734131
  29:     0x7fd0ca988a70 - sys::thread::thread_start::h15e840368764388fGdH
  30:     0x7fd0c49cda80 - start_thread
  31:                0x0 - <unknown>

Could not compile `android_glue`.

To learn more, run the command again with --verbose.
installing application
```

### LINKED ISSUE #23151: Compiler crash when matching generic enum
<short summary of the bug>

I tried this code:

```rust
enum Foo<T> {
    A(T),
    B
}

fn main() {
    let x = Foo::<u32>::A(14);
    let y = match x {
        Foo<T>::A(value) => value,
        Foo<T>::B => 7,
    };
    println!("{}", y);
}
```
and it gave me this message:

```
error.rs:9:9: 9:12 error: internal compiler error: ident only path should have been covered already
error.rs:9         Foo<T>::A(value) => value,
                   ^~~
note: the compiler unexpectedly panicked. this is a bug.
note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports
note: run with `RUST_BACKTRACE=1` for a backtrace
thread 'rustc' panicked at 'Box<Any>', /tmp/portage/dev-lang/rust-1.0.0_alpha2/work/rustc-1.0.0-alpha.2/src/libsyntax/diagnostic.rs:129
```

I have git version of rust from package in funtoo linux.
When I tried the code at play.rust-lang.org the example produced compiler error, but not a crash.

## Meta

`rustc --version-verbose`:
```
rustc 1.0.0-dev (built 2015-02-28)
binary: rustc
commit-hash: unknown
commit-date: unknown
build-date: 2015-02-28
host: x86_64-unknown-linux-gnu
release: 1.0.0-dev
```

Backtrace:
```
   1:     0x7f894f434c20 - sys::backtrace::write::h6db1343660ca93a2KlC
   2:     0x7f894f45a900 - <unknown>
   3:     0x7f894f3a5d80 - rt::unwind::begin_unwind_inner::h734e583390472682RBL
   4:     0x7f894c8fa4e0 - <unknown>
   5:     0x7f894c8fa470 - diagnostic::SpanHandler::span_bug::ha4389e9b9cd26b0cvYE
   6:     0x7f894c91c950 - parse::parser::Parser<'a>::parse_pat::hbaca44ef1986f811wLL
   7:     0x7f894c9314c0 - parse::parser::Parser<'a>::parse_arm::h03dad4fde878481bYzL
   8:     0x7f894c926950 - parse::parser::Parser<'a>::parse_bottom_expr::hd51982e53d14adeaKwK
   9:     0x7f894c92c9a0 - parse::parser::Parser<'a>::parse_dot_or_call_expr::h73cc376278c09c13LPK
  10:     0x7f894c92f7c0 - parse::parser::Parser<'a>::parse_prefix_expr::h9ada9bffadff12ebpcL
  11:     0x7f894c930020 - parse::parser::Parser<'a>::parse_binops::hd3006072d5866dc6shL
  12:     0x7f894c930600 - parse::parser::Parser<'a>::parse_assign_expr::hb68dbae6fdb97b62imL
  13:     0x7f894c932420 - <unknown>
  14:     0x7f894c932640 - parse::parser::Parser<'a>::parse_stmt::h3a7a8bca0f5459e8v4L
  15:     0x7f894c93bff0 - <unknown>
  16:     0x7f894c91add0 - <unknown>
  17:     0x7f894c940630 - <unknown>
  18:     0x7f894c9352c0 - <unknown>
  19:     0x7f894c946830 - <unknown>
  20:     0x7f894c951b90 - parse::parser::Parser<'a>::parse_crate_mod::h1592f06e4a607835FlO
  21:     0x7f894c963d90 - parse::parse_crate_from_file::h71bb673ba4ce988fvNV
  22:     0x7f894fa4f920 - <unknown>
  23:     0x7f894fa29c10 - driver::phase_1_parse_input::h5a91fba8b70d831dQra
  24:     0x7f894fa26610 - driver::compile_input::ha5d27daf21872ba2Gba
  25:     0x7f894faf4b50 - run_compiler::he50deb5aedf688fcZbc
  26:     0x7f894faf3480 - <unknown>
  27:     0x7f894faf2370 - <unknown>
  28:     0x7f894f4b7bc0 - <unknown>
  29:     0x7f894f4b7bb0 - rust_try
  30:     0x7f894faf2670 - <unknown>
  31:     0x7f894f448650 - <unknown>
  32:     0x7f894a0acf40 - <unknown>
  33:     0x7f894f02b7d9 - clone
  34:                0x0 - <unknown>
```

### LINKED ISSUE #23597: ICE: ident only path should have been covered already
Minimal example to reproduce (malformed array declaration): 

```
fn main() {
    let bla[isize;5] = [0;5];
}
```

Output
```
src/main.rs:2:9: 2:12 error: internal compiler error: ident only path should have been covered already
src/main.rs:2     let bla[isize;5] = [0;5];
                      ^~~
note: the compiler unexpectedly panicked. this is a bug.
note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports
note: run with `RUST_BACKTRACE=1` for a backtrace
thread 'rustc' panicked at 'Box<Any>', /Users/rustbuild/src/rust-buildbot/slave/nightly-dist-rustc-mac/build/src/libsyntax/diagnostic.rs:130

stack backtrace:
   1:        0x11119625b - sys::backtrace::write::hf72be5f813c90d15u5C
   2:        0x1111c1918 - panicking::on_panic::h357b9ffd8bc8f79dSWI
   3:        0x1110de7f7 - rt::unwind::begin_unwind_inner::h12b438ebbd8042edUFI
   4:        0x1108d37ce - rt::unwind::begin_unwind::h4369435781297418514
   5:        0x1108d377b - diagnostic::SpanHandler::span_bug::h853a022b690aedcaDaB
   6:        0x1108fbe58 - parse::parser::Parser<'a>::parse_pat::hc63de28bd64b0b52oBH
   7:        0x11090ff01 - parse::parser::Parser<'a>::parse_local::h9c98c8f6a10fa0c3ISH
   8:        0x110910930 - parse::parser::Parser<'a>::parse_stmt_::h89dfb0c159e79c82PVH
   9:        0x1109090eb - parse::parser::Parser<'a>::parse_block_tail::h29764a009812f1c395H
  10:        0x1108f5b9f - parse::parser::Parser<'a>::parse_inner_attrs_and_block::hbcd241202ba69a8024H
  11:        0x11091c93f - parse::parser::Parser<'a>::parse_item_fn::h07befed94e01b7afSHI
  12:        0x1109137e9 - parse::parser::Parser<'a>::parse_item_::h1a7567183f075285hCJ
  13:        0x110922781 - parse::parser::Parser<'a>::parse_mod_items::h34bd4c448262f56bR5I
  14:        0x1109287fa - parse::parser::Parser<'a>::parse_crate_mod::h66c5e5883ab0fe0d50J
  15:        0x11093d630 - parse::parse_crate_from_file::hce10797f00bc1302SPQ
  16:        0x10d6f85ee - driver::phase_1_parse_input::closure.15455
  17:        0x10d6d204c - driver::phase_1_parse_input::h8843c0cb78a2306dOqa
  18:        0x10d6cf0ef - driver::compile_input::h7fdf739785f6d9d4Rba
  19:        0x10d7907d3 - run_compiler::hb484336e43423a0fv2b
  20:        0x10d78e2f5 - thunk::F.Invoke<A, R>::invoke::h11654417678186204529
  21:        0x10d78d6b7 - rt::unwind::try::try_fn::h10741634765121132143
  22:        0x111245f68 - rust_try_inner
  23:        0x111245f55 - rust_try
  24:        0x10d78da55 - thunk::F.Invoke<A, R>::invoke::h13457747256212984027
  25:        0x1111ac83d - sys::thread::create::thread_start::h937519545136f3ecJBH
  26:     0x7fff94a06267 - _pthread_body
  27:     0x7fff94a061e4 - _pthread_start

Could not compile `test-rs`.

To learn more, run the command again with --verbose.
```

Rust version:
```
rustc --version                                                                              
rustc 1.0.0-nightly (f4e0ce66a 2015-03-19) (built 2015-03-20)
```

### LINKED ISSUE #23656: internal compiler error by using a string slice in pattern match
Attempting to get a string slice in a pattern when pattern matching yields the following: "error: internal compiler error: ident only path should have been covered already." As an example, do consider the following code:
http://is.gd/cuYVtu

```rust
struct Test(String);

fn main() {
    let x = Test(format!("foo"));
    let Test(&desc[..]) = x;
}
```

I expected a less serious error, as this was by accident, I intended to do something different. The version I am using is: rustc 1.0.0-dev (ecf8c64e1 2015-03-21) (built 2015-03-22).

### LINKED ISSUE #23929: ICE with invalid for loop: "ident only path should have been covered already"
Code: 

```rust
fn main() {
	for thing(x[]) {
	}
}
```

Error:

```
<anon>:2:12: 2:13 error: internal compiler error: ident only path should have been covered already
<anon>:2 	for thing(x[]) {
         	          ^
note: the compiler unexpectedly panicked. this is a bug.
note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports
note: run with `RUST_BACKTRACE=1` for a backtrace
thread 'rustc' panicked at 'Box<Any>', /home/rustbuild/src/rust-buildbot/slave/nightly-dist-rustc-linux/build/src/libsyntax/diagnostic.rs:130
```

[playpen](http://is.gd/U7Jy7A)

### PR DESCRIPTION
**Pull request by petrochenkov**
Fixes #22757
Fixes #22972
Fixes #23044
Fixes #23151
Fixes #23597
Fixes #23656
Fixes #23929
It also fixes some other corner cases in range patterns, like incorrect spans or not accepting global paths after `...`.

It passes `make check` but needs some additional tests (then it will fix #22546 as well), I'll write them today or tomorrow.

### CODE REVIEW

**Comment by alexcrichton** [comment]
Nice!

File: src/libsyntax/parse/parser.rs [hunk 1]
```diff
@@ -2603,13 +2601,6 @@ impl<'a> Parser<'a> {
     pub fn parse_more_binops(&mut self, lhs: P<Expr>, min_prec: usize) -> P<Expr> {
         if self.expr_is_complete(&*lhs) { return lhs; }
 
-        // Prevent dynamic borrow errors later on by limiting the
-        // scope of the borrows.
-        if self.token == token::BinOp(token::Or) &&
-            self.restrictions.contains(RESTRICTION_NO_BAR_OP) {
-            return lhs;
-        }
```
Comment (nrc) [inline src/libsyntax/parse/parser.rs]: Could you explain the reasoning behind why it is safe to remove this check please?

**Comment by nrc** [comment]
lgtm, I just want to be clear about removing that one check

File: src/libsyntax/parse/parser.rs (hunk 1 above)
Comment (petrochenkov) [inline src/libsyntax/parse/parser.rs]: Flag RESTRICTION_NO_BAR_OP is checked here but it isn't set anywhere, so I removed the check and the flag itself. RESTRICTION_NO_BAR_OP was used before in `parse_pat`, see my line comments there.

File: src/libsyntax/parse/parser.rs [hunk 2]
```diff
@@ -3087,250 +3094,145 @@ impl<'a> Parser<'a> {
...
-                    token::OpenDelim(_) | token::Lt | token::ModSep => true,
-                    _ => false,
-                }
-            });
-
-            if self.look_ahead(1, |t| *t == token::DotDotDot) &&
-                    self.look_ahead(2, |t| {
-                        *t != token::Comma && *t != token::CloseDelim(token::Bracket)
-                    }) {
-                let start = self.parse_expr_res(RESTRICTION_NO_BAR_OP);
```
Comment (petrochenkov) [inline src/libsyntax/parse/parser.rs]: Here `parse_expr_res(RESTRICTION_NO_BAR_OP)` was mistakenly used to parse plain identifier.

File: src/libsyntax/parse/parser.rs [hunk 3]
```diff
@@ -3087,250 +3094,145 @@ impl<'a> Parser<'a> {
...
-                self.eat(&token::DotDotDot);
-                let end = self.parse_expr_res(RESTRICTION_NO_BAR_OP);
```
Comment (petrochenkov) [inline src/libsyntax/parse/parser.rs]: This is the only place where end of range pattern was parsed as a general expression with `parse_expr_res(RESTRICTION_NO_BAR_OP)` and not as path. I'm almost sure it was a mistake.

**Comment by nrc** [comment]
Ok, sounds good, thanks for the explanation.

@bors: r+ d9a0ca01f3938efbb34587e3f52214d8c0bc1624

**Comment by petrochenkov** [comment]
Rebased and fixed.

**Comment by jakub-** [comment]
@petrochenkov Did you mean to say that you'd add the test cases in a separate PR? As it is, none of the fixed issues have any related tests so perhaps we should reopen them?

**Comment by petrochenkov** [comment]
@jakub- 
I meant to write the tests as a part of this PR but was busy, so I'm going to submit them separately. All the mentioned issues are essentially duplicates, so I'll open one blanket E-needstest issue for them if I don't submit the PR with tests in the next couple of hours.

**Comment by bombless** [comment]
Hmm @bors closed https://github.com/rust-lang/rust/issues/22546 after merging this, but I believe it is an irrelevant issue, since the core idea of that issue is about discussing whether we should allow type parameters in pattern.

**Comment by bombless** [comment]
cc @pnkfelix

**Comment by brson** [comment]
@bombless I reopened that issue for you.

**Comment by brson** [comment]
Awesome list of fixes.

**Comment by pnkfelix** [comment]
@bombless thanks for the pro-active note.  However, during triage (I think), the team decided as noted in e.g. [this comment](https://github.com/rust-lang/rust/issues/22546#issuecomment-76277026) that we would keep the syntax, and that we just needed to exercise the edge cases more thoroughly.

**Comment by pnkfelix** [comment]
But, skimming this PR, it seems like it did not actually add any tests relevant to #22546, so I guess that issue should remain open.

**Comment by petrochenkov** [comment]
@pnkfelix 
The follow-up PR (#24033) contains some tests for #22546. I don't know, whether that enough or not.

**Pull request merged**